### PR TITLE
ArC: fix some scenarios with generic decorators

### LINKED ISSUE #30811: CDI Decorator ignored for beans implementing interface extending other generic interface
### Describe the bug

CDI Decorator is ignored for beans implementing interface extending other generic interface. 
E.g.
```
public interface MyInterface extends Consumer<String> {
}
```

IMHO is the bug in [BeanInfo.findMatchingDecorators](https://github.com/quarkusio/quarkus/blob/1c12f86c16ffa5dbd4b6e74b67b0c7d44744d9b1/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanInfo.java#L659)

In this case method name is matched on Consumer interface (`void accept(T t)`), but because `Consumer` is generic, its parameter type (`T`) cannot be matched against the parameter type of the implementation (`String`).

Repeating the parent interface method declaration fixes this issue e.g.:

```
public interface MyInterface extends Consumer<String> {
   void accept(String t);
}
```

### Expected behavior

_No response_

### Actual behavior

_No response_

### How to Reproduce?

```
import java.util.function.Consumer;

import javax.annotation.Priority;
import javax.decorator.Decorator;
import javax.decorator.Delegate;
import javax.enterprise.context.Dependent;
import javax.enterprise.inject.Any;
import javax.inject.Inject;

import io.quarkus.runtime.QuarkusApplication;
import io.quarkus.runtime.annotations.QuarkusMain;

@QuarkusMain
public class DecoratorTest implements QuarkusApplication {

    public interface NotWorking extends Consumer<String> {

    }

    @Dependent
    public static class NotWorkingDefaultImpl implements NotWorking {

        @Override
        public void accept(String t) {
            System.out.println("NotWorkingDefaultImpl " + t);
        }

    }

    @Priority(1)
    @Decorator
    public static class NotWorkingDecorator implements NotWorking {

        @Inject
        @Delegate
        @Any
        NotWorking delegate;

        @Override
        public void accept(String t) {
            System.out.println("NotWorkingDecorator " + t);
            delegate.accept(t);

        }

    }

    public interface Working extends Consumer<String> {
        // must redeclare accept method from Consumer for decorator to work
        @Override
        void accept(String t);
    }

    @Dependent
    public static class WorkingDefaultImpl implements Working {

        @Override
        public void accept(String t) {
            System.out.println("WorkingDefaultImpl " + t);
        }

    }

    @Priority(1)
    @Decorator
    public static class WorkingDecorator implements Working {

        @Inject
        @Delegate
        @Any
        Working delegate;

        @Override
        public void accept(String t) {
            System.out.println("WorkingDecorator " + t);
            delegate.accept(t);

        }

    }

    @Inject
    NotWorking decoratorNotWorking;

    @Inject
    Working decoratorWorking;

    @Override
    public int run(String... args) throws Exception {

        /*
         * prints:
         * 
         * NotWorkingDefaultImpl value
         */
        decoratorNotWorking.accept("value");

        /*
         * prints:
         * 
         * WorkingDecorator value
         * 
         * WorkingDefaultImpl value
         */
        decoratorWorking.accept("value");
        return 0;
    }

}


```

### Output of `uname -a` or `ver`

_No response_

### Output of `java -version`

_No response_

### GraalVM version (if different from Java)

_No response_

### Quarkus version or git rev

_No response_

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

_No response_

### Additional information

_No response_

### LINKED ISSUE #33803: Generic CDI Decorator Throws java.lang.NoSuchMethodError
### Describe the bug

When using CDI decorators with generic type parameters in a Quarkus application, a `NoSuchMethodError` is thrown when invoking a decorated method. The error message indicates that it cannot find the method in the decorator class.

The following configuration fails when executing the `GreetingResource.hello()` method:

```java
interface Repository<T> {
  T get();
}

@Priority(APPLICATION)
@Decorator
class LoggingRepository<T> implements Repository<T> {

  @Inject Logger logger;

  @Inject @Any @Delegate Repository<T> delegate;

  @Override
  public T get() {
    logger.debug("Getting item");
    return delegate.get();
  }
}

@ApplicationScoped
class GreetingRepository implements Repository<String> {

  @Override
  public String get() {
    return "Hello from RESTEasy Reactive";
  }
}

@Path("/hello")
class GreetingResource {

  @Inject GreetingRepository repository;

  @GET
  @Produces(MediaType.TEXT_PLAIN)
  public String hello() {
    return repository.get();
  }
}
```

The exception is:

```
	Suppressed: java.lang.NoSuchMethodError: 'java.lang.String org.acme.LoggingRepository.get()'
		at org.acme.GreetingRepository_Subclass.get(Unknown Source)
		at org.acme.GreetingRepository_ClientProxy.get(Unknown Source)
		at org.acme.GreetingResource.hello(GreetingResource.java:53)
		at org.acme.GreetingResource$quarkusrestinvoker$hello_e747664148511e1e5212d3e0f4b40d45c56ab8a1.invoke(Unknown Source)
		at org.jboss.resteasy.reactive.server.handlers.InvocationHandler.handle(InvocationHandler.java:29)
		at io.quarkus.resteasy.reactive.server.runtime.QuarkusResteasyReactiveRequestContext.invokeHandler(QuarkusResteasyReactiveRequestContext.java:141)
		at org.jboss.resteasy.reactive.common.core.AbstractResteasyReactiveContext.run(AbstractResteasyReactiveContext.java:145)
```

A similar configuration works as expected in Weld.

### Expected behavior

Quarkus Arc decorates all beans that implement the interface with the decorator and no exception is thrown. In this example, the `GreetingRepository` is decorated with logic in the `LoggingRepository<T>` decorator.

### Actual behavior

Decorated bean throws `NoSuchMethodException`.

### How to Reproduce?

[code-with-quarkus.zip](https://github.com/quarkusio/quarkus/files/11638777/code-with-quarkus.zip)

### Output of `uname -a` or `ver`

Darwin pixee-mbp-gilday.localdomain 22.4.0 Darwin Kernel Version 22.4.0: Mon Mar  6 20:59:28 PST 2023; root:xnu-8796.101.5~3/RELEASE_ARM64_T6000 arm64

### Output of `java -version`

openjdk version "17.0.7" 2023-04-18 OpenJDK Runtime Environment Temurin-17.0.7+7 (build 17.0.7+7) OpenJDK 64-Bit Server VM Temurin-17.0.7+7 (build 17.0.7+7, mixed mode)

### GraalVM version (if different from Java)

_No response_

### Quarkus version or git rev

3.1.0.Final

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Apache Maven 3.8.8 (4c87b05d9aedce574290d1acc98575ed5eb6cd39) Maven home: /Users/jgilday/.m2/wrapper/dists/apache-maven-3.8.8-bin/67c30f74/apache-maven-3.8.8 Java version: 17.0.7, vendor: Eclipse Adoptium, runtime: /Library/Java/JavaVirtualMachines/temurin-17.jdk/Contents/Home Default locale: en_US, platform encoding: UTF-8 OS name: "mac os x", version: "13.3.1", arch: "aarch64", family: "mac"

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by mkouba**
- fixes #33803
- fixes #30811

### CODE REVIEW

**Comment by mkouba** [comment]
This should also fix [#30811](https://github.com/quarkusio/quarkus/issues/30811).

**Review by manovotn — COMMENTED** [review]
I like the `DecoratorMethod` as means of carrying over the decorator's method signature :+1: 
Curious to see if CI shows any problems but I doubt we have much decorator usage in code.

File: independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanInfo.java [hunk 1]
```diff
@@ -750,13 +769,20 @@ private List<DecoratorInfo> findMatchingDecorators(MethodInfo method, List<Decor
                             decoratedMethod,
                             beanDeployment.getBeanArchiveIndex());
                     for (int i = 0; i < methodParams.size(); i++) {
-                        if (!beanDeployment.getDelegateInjectionPointResolver().matches(decoratedMethodParams.get(i),
-                                methodParams.get(i))) {
-                            matches = false;
+                        BeanResolver resolver = beanDeployment.getDelegateInjectionPointResolver();
+                        Type decoratedParam = decoratedMethodParams.get(i);
+                        if (decoratedParam.kind() == org.jboss.jandex.Type.Kind.TYPE_VARIABLE) {
+                            if (!resolver.matchTypeArguments(decoratedMethodParams.get(i), methodParams.get(i))) {
```
Comment (Ladicek) [inline independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanInfo.java:775]:
```suggestion
                            if (!resolver.matchTypeArguments(decoratedParam, methodParams.get(i))) {
```
?

File: independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanInfo.java [hunk 2]
```diff
@@ -750,13 +769,20 @@ private List<DecoratorInfo> findMatchingDecorators(MethodInfo method, List<Decor
...
+                                matches = false;
+                            }
+                        } else {
+                            if (!resolver.matches(decoratedMethodParams.get(i), methodParams.get(i))) {
```
Comment (Ladicek) [inline independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanInfo.java:779]:
```suggestion
                            if (!resolver.matches(decoratedParam, methodParams.get(i))) {
```
?

File: independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/SubclassGenerator.java [hunk 3]
```diff
@@ -617,9 +623,12 @@ private void processDecorator(DecoratorInfo decorator, BeanInfo bean, Type provi
         }
         ClassCreator delegateSubclass = delegateSubclassBuilder.build();
 
-        Map<MethodDescriptor, DecoratorInfo> nextDecorators = bean.getNextDecorators(decorator);
-        Collection<DecoratorInfo> nextDecoratorsValues = nextDecorators.values();
-        List<DecoratorInfo> decoratorParameters = new ArrayList<>(new HashSet<>(nextDecoratorsValues));
+        Map<MethodDescriptor, DecoratorMethod> nextDecorators = bean.getNextDecorators(decorator);
+        Set<DecoratorInfo> nextDecoratorsValues = new HashSet<>();
```
Comment (Ladicek) [inline independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/SubclassGenerator.java:627]: The set here is used to eliminate possible duplicates? Just for my understanding.

File: independent-projects/arc/tests/src/test/java/io/quarkus/arc/test/decorators/generics/MultipleDecoratorsWithTypeVariablesTest.java [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+
+        @SuppressWarnings("unchecked")
+        @Override
+        public T apply(String val) {
+            return (T) delegate.apply(val).toString().repeat(2);
+        }
+
+    }
+
+    @Priority(3)
```
Comment (Ladicek) [inline independent-projects/arc/tests/src/test/java/io/quarkus/arc/test/decorators/generics/MultipleDecoratorsWithTypeVariablesTest.java:95]:
```suggestion
    @Priority(2)
```

Doesn't make a difference, but would be more consistent.

**Review by Ladicek — APPROVED** [review]
LGTM, though I admit I only skimmed through the bytecode generation part, I don't know the decorator-related parts of `SubclassGenerator` very well.

File: independent-projects/arc/tests/src/test/java/io/quarkus/arc/test/decorators/generics/MultipleDecoratorsWithTypeVariablesTest.java (hunk 4 above)
Comment (mkouba) [reply independent-projects/arc/tests/src/test/java/io/quarkus/arc/test/decorators/generics/MultipleDecoratorsWithTypeVariablesTest.java:95]: Hm, that was intentional - tests do not need to be consistent... I wanted to break the sequence.

Comment (Ladicek) [reply independent-projects/arc/tests/src/test/java/io/quarkus/arc/test/decorators/generics/MultipleDecoratorsWithTypeVariablesTest.java:95]: I agree tests don't have to be consistent -- but _all_ of them added in this PR are. That's how this caught my attention.

**Review by manovotn — APPROVED** [review]

**Pull request merged**